packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -177,3 +177,159 @@ class DocumentBoundary extends TextBoundary {
     );
   }
 }
+
+/// A text boundary that use the first non-whitespace character as the logical
+/// boundary.
+///
+/// This text boundary uses [TextLayoutMetrics.isWhitespace] to identify white
+/// spaces, this include newline characters from ASCII and separators from the
+/// [unicode separator category](https://www.compart.com/en/unicode/category/Zs).
+class WhitespaceBoundary extends TextBoundary {
+  /// Creates a [WhitespaceBoundary] with the text.
+  const WhitespaceBoundary(this._text);
+
+  final String _text;
+
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    // Handles outside of right bound.
+    if (position.offset > _text.length || (position.offset == _text.length  && position.affinity == TextAffinity.downstream)) {
+      position = TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+    }
+    // Handles outside of left bound.
+    if (position.offset <= 0) {
+      return const TextPosition(offset: 0);
+    }
+    int index = position.offset;
+    if (position.affinity == TextAffinity.downstream && !TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index))) {
+      return position;
+    }
+
+    for (index -= 1; index >= 0; index -= 1) {
```

Comment:
I think it would be easier to read if there's an _effectiveTextOffset method that turns caret locations to `int` offsets?

Reply:
while loop makes sense, I am not sure how _effectiveTextOffset should be used to implement this for loop. Can you elaborate?

Reply:
Something like (I thought you added a similar function in the previous PR?):
`int _toTextOffset(TextPosition position) => max(0, min(upstream ? position.offset - 1 : position.offset, textLength - 1));`
Then:
```dart
int offset = _toTextOffset(position);
while (offset >= 0 && !TextLayoutMetrics.isWhitespace(_text.codeUnitAt(offset)) { 
  offset -= 1;
}
return Offset(offset: offset + 1, affinity: offset < 0 ? upstream : downstream); // Is the affinity correct?
```

Reply:
The reason I don't want to change the the position to text offset before the first
```
    if (position.affinity == TextAffinity.downstream && !TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index))) {
      return position;
    }
    ```
is that I need to preserve the affinity of the input if the input already point to a non whitespace character.

An example lets say you have string without any white space 'abcd'.

The expected behavior is the following:
getLeadingTextBoundaryAt(TextPosition(3, upstream)) = TextPosition(3, upstream)
getLeadingTextBoundaryAt(TextPosition(2, downstream)) = TextPosition(2, downstream)

however your code will produce

getLeadingTextBoundaryAt(TextPosition(3, upstream)) = TextPosition(3, upstream)
getLeadingTextBoundaryAt(TextPosition(2, downstream)) = TextPosition(3, upstream)

The problem is you can't tell between `TextPosition(3, upstream)` and `TextPosition(2, downstream)` because they are the same in text offset. If you convert it first before the return, you can't differentiate  them.

I could do the conversion after the first return, but at that point, I don't care about affinity anymore because I only need to find first non whitespace character and do (index+1 upstream).